Upmerge branch

Vge0rge opened the pull request:
This uses MbedTLS 3.0 as a base and it has some downstream changes and the changes that the mbedtls fork of zephyr uses.

tamasban committed:
mbedtls: remove static keyword from certain function pointers

TF-M usually downloads mbedtls and then applies a handful of
patches. These patches were forgotten when Zephyr started to use it's
own mbedtls fork with TF-M.

One of these patches has now been re-applied in this commit. It is
needed for an important code sharing optimization documented in a link
below.

The static keyword is removed from three function pointer
declarations. Aside from allowing the optimization described below
this could theoretically backfire.

There could be other non-static symbols that now trigger aliasing
issues.

The compiler could have trouble optimizing now that it doesn't know of
all symbol usages.

But I believe that in practice this will be OK.

I don't know if there are any plans to upstream this patch to mbedtls
itself.

https://tf-m-user-guide.trustedfirmware.org/docs/technical_references/design_docs/code_sharing.html

Original patch:
https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/commit/lib/ext/mbedcrypto?id=4a5cc9776ed3b053bac57326931f936fbbc660e9

Signed-off-by: Tamas Ban <tamas.ban@arm.com>
Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

Summer-ARM committed:
Disable export MbedTLSTargets

Disable install MbedTLSConfig.cmake, MbedTLSConfigVersion.cmake and
MbedTLSTargets.cmake. And Disable export MbedTLSTargets since this
needs to configure some tfm target which for mbedtls to be exported.

Signed-off-by: Summer Qin <summer.qin@arm.com>
[david.brown: Copied from https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/tree/lib/ext/mbedcrypto/0003-Disable-export-MbedTLSTargets.patch]
Signed-off-by: David Brown <david.brown@linaro.org>

Commit:
[nrf]Add driver_wrapper call for key agreement

-This forwards the psa call for raw_key_agreement to the relevant
 PSA driver wrapper function in order to allow usage of drivers

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

d3zd3z committed:
west: Add zephyr module

Bring the zephyr module file from the previous branch, making this tree
a valid Zephyr module.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
Merge pull request #31 from d3zd3z/tozephyr

Apply Zephyr-specific changes to upstream

urutva committed:
Disable dependency check for `MBEDTLS_USE_PSA_CRYPTO` on `MBEDTLS_PSA_CRYPTO_C`

The config option `MBEDTLS_USE_PSA_CRYPTO` enables PSA/TF-M backend for
MbedTLS. But there is a dependency check for this config option on
`MBEDTLS_PSA_CRYPTO_C` which enables PSA crypto implementation in
MbedTLS.

Using this config option in Zephyr leads to linker error due to multiple
definition of PSA APIs (TF-M and MbedTLS PSA Crypto). For detailed
description check https://github.com/ARMmbed/mbedtls/issues/4597.

Temporarily disable this check till the aforementioned issue is resolved
upstream.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@linaro.org>

d3zd3z committed:
Merge pull request #33 from urutva/disable_dependency_check

Disable dependency check for `MBEDTLS_USE_PSA_CRYPTO` on `MBEDTLS_PSA_CRYPTO_C`

frkv committed:
[nrf]Undef PSA crypto client in crypto_types

-To remove unnecessary build warnings for define being redefined
 we need to undefine the MBEDTLS_PSA_CRYPTO_CLIENT in the
 crypto_types header file

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>

frkv committed:
[nrf]Disable turning of legacy context for ECDH

-This enables using the ECDH legacy context

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>

Vge0rge committed:
[nrf]Remove redefinition ECC_MAX_CURVE_BITS

-This removes the redefinition of the define
 PSA_VENDOR_ECC_MAX_CURVE_BITS which we will set
 in our configuration file.

 Ref: NCSDK-12898

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

Vge0rge committed:
[nrf]Add AEAD PSA multipart support

-This adds AEAD multipart support for the
 PSA APIs

Ref: NCSDK-13181

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

Vge0rge committed:
[nrf]Add CC3XX driver includes

-This adds the include headers needed for
 CC3XX driver support

Ref: NCSDK-13181

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

SebastianBoe reviewed (changes requested):
We need to re-work this history.

We should be forking Zephyr's mbedtls instead of ARM's mbedtls.

This will get rid of some of the commits here.

Also, the sauce tags are badly formatted, should be [nrf noup] etc.

Vge0rge committed:
[nrf]Check if ECP_MAX_BITS is set in ecp.h

-The ECP_ALT file sets this variable so we
  will have a multiple definition if we enable
  it in ecp.h.

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

Vge0rge committed:
[nrf]Check if GCM_C is defined in gcm.h

-This checks if GCM_C is enabled in gcm.h
 before including the functions. This was
 causing build issues when the GCM is disabled
 but GCM_ALT is enabled.

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>

LuDuda commented on include/mbedtls/gcm.h:
```diff
@@ -375,5 +377,6 @@ int mbedtls_gcm_self_test( int verbose );
 }
 #endif
 
+#endif /* MBEDTLS_GCM_C */
```
My C++ code does not compile and throws mysterious errors :smile: Figured out that this is the reason.. this line should be in line 375.

Else we have:

``
extern "C" {
``

that is not closed when MBEDTLS_GCM_C is not defined.

Vge0rge replied:
Should be done :)

frkv replied:
Should be fixed with a force-push

SebastianBoe commented:
a variant of this, that had slightly different git commit messages, was force pushed to the development branch.
